File: exercise/40/solution.cpp
class Solution {
public:
    void dfs (vector<vector<int>> &result, vector<int> &now, int res, vector<pair<int, int>> &nums, int index) {
        if (res == 0) {
            result.push_back(now);
            return;
        }
        if (nums[index].first > res) return;
        if (index + 1 < nums.size()) {
            dfs(result, now, res, nums, index + 1);
        }
        if (nums[index].second > 0) {
            now.push_back(nums[index].first);
            nums[index].second--;
            dfs(result, now, res - nums[index].first, nums, index);
            nums[index].second++;
            now.pop_back();
        }
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        if (candidates.empty()) return result;
        sort(candidates.begin(), candidates.end());
        vector<pair<int, int>> nums;
        nums.push_back(make_pair(candidates[0], 1));
        for (int i = 1; i < candidates.size(); i++) {
            if (candidates[i] == nums.back().first) {
                nums.back().second++;
            } else {
                nums.push_back(make_pair(candidates[i], 1));
            }
        }
        vector<int> now;
        dfs(result, now, target, nums, 0);
        return result;
    }
};

File: exercise/40/solution.py
class Solution:
    def combinationSum2Part(self, candidates, target, index, result, now):
        if target == 0:
            result.append(list(now))
        elif index >= len(candidates) or candidates[index] > target:
            return
        else:
            now.append(candidates[index])
            self.combinationSum2Part(candidates, target - candidates[index], index + 1, result, now)
            now.pop()
            next_index = index + 1
            while next_index < len(candidates) and candidates[next_index] == candidates[next_index-1]:
                next_index += 1
            self.combinationSum2Part(candidates, target, next_index, result, now)

    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        result = []
        now = []
        self.combinationSum2Part(candidates, target, 0, result, now)
        return result

